sklearn/decomposition/nmf.py

```diff
@@ -254,8 +340,222 @@ def _nls_subproblem(V, W, H, tol, max_iter, sigma=0.01, beta=0.1):
+        W, _, _ = _update_projected_gradient_w(X, W, H, tol, nls_max_iter,
+                                               alpha, l1_ratio, sparseness,
+                                               beta, eta)
+
+    return W, H, n_iter
+
+
+def _update_coordinate_descent(X, W, Ht, alpha, l1_ratio):
+    """Helper function for fit_coordinate_descent"""
+    WtW = fast_dot(W.T, W)
+    WtX = safe_sparse_dot(W.T, X)
+
+    if WtW.flags['F_CONTIGUOUS']:
+        WtW = WtW.T
+
+    if WtX.flags['F_CONTIGUOUS']:
+        return _update_cdnmf_fast(Ht, WtW, WtX.T, alpha, l1_ratio, True)
+    else:
+        return _update_cdnmf_fast(Ht, WtW, WtX, alpha, l1_ratio, False)
+
+
+def fit_coordinate_descent(X, W, H, tol=1e-4, max_iter=200, alpha=0.001,
+                           l1_ratio=0., regularization=None,
+                           update_H=True, verbose=0):
+    """Compute Non-negative Matrix Factorization (NMF) with Coordinate Descent
+
+    Find (W, H) that minimize || X - W.H ||_2
+    with an alternative least square minimization of W and H.
+    Each minimization is done with a Coordinate Descent.
+
+    Parameters
+    ----------
+    X : array-like, shape (n_samples, n_features)
+        Constant matrix.
+
+    W : array-like, shape (n_samples, n_components)
+        Initial guess for the solution.
+
+    H : array-like, shape (n_components, n_features)
+        Initial guess for the solution.
+
+    tol : float, default: 1e-4
+        Tolerance of the stopping condition.
+
+    max_iter : integer, default: 200
+        Maximum number of iterations before timing out.
+
+    alpha : double, default: 0.
+        Constant that multiplies the regularization terms.
+
+    l1_ratio : double, default: 0.
+        The regularization mixing parameter, with 0 <= l1_ratio <= 1.
+        For l1_ratio = 0 the penalty is an L2 penalty.
+        For l1_ratio = 1 it is an L1 penalty.
+        For 0 < l1_ratio < 1, the penalty is a combination of L1 and L2.
+
+    regularization : 'both' | 'components' | 'transformation' | None
+        Select whether the regularization affects the components (H), the
+        transformation (W), both or none of them.
+
+    update_H : boolean, default: True
+        Set to True, both W and H will be estimated from initial guesses.
+        Set to False, only W will be estimated.
+
+    verbose : integer, default: 0
+        The verbosity level.
+
+    Returns
+    -------
+    W : array-like, shape (n_samples, n_components)
+        Solution to the non-negative least squares problem.
+
+    H : array-like, shape (n_components, n_features)
+        Solution to the non-negative least squares problem.
+
+    n_iter : int
+        The number of iterations done by the algorithm.
+
+    References
+    ----------
+    Hsieh, Cho-Jui, and Inderjit S. Dhillon. "Fast coordinate descent methods
+    with variable selection for non-negative matrix factorization."
+    Proceedings of the 17th ACM SIGKDD. ACM, 2011.
```

Comment:
Thanks for the note, that's what it seemed to me to be the case as well. If our CD implementation will only support the Frobenius norm, then I think it would be worth it to try the greedy selection. If I understood the paper correctly, the trick cannot be efficiently applied for I-divergence, right?